### DISCUSSION TITLE
How to specify the GeoJSON top-level "name" field when writing a file?

### DISCUSSION THREAD

**Discussion by nk9** [Toblerity/Fiona#1348] [Q&A]
I have code to write an array of features out to a file. This is working, except that I want the generated GeoJSON file to have a specific top-level `name` attribute which is not derived from the file's name. Right now, the `name` is not included when Fiona writes the file. However, when a file is edied with QGIS, it adds the `name` if it's not present, defaulting to the file name without the `.geojson` suffix.

I've tried passing the string I want as the `layer` in the `open()` command, but that just results in "[the GeoJSON format does not have layers](https://github.com/Toblerity/Fiona/blob/42af94db4d7c8944c0b7bb736b8b6776655de8a9/fiona/collection.py#L181C39-L181C78)". How do I configure this?

**Explanation:** I'm using this field because it allows me to adjust the default layer name when the file is loaded into QGIS. I have a bunch of files, all of which are named "eds.geojson". They are identified by their location in a directory structure. But when loaded into QGIS, I want the added layer to be named "eds — my supplied layer name". And, as above: if I don't provide the name, QGIS will add it for me.

**Comment by jmarintur** [comment]
Hi @nk9, fiona does not support naming geojsons, and more importantly, geojsons do not include a provision for layer names (see [here](https://datatracker.ietf.org/doc/html/rfc7946)). That said, I'd suggest you to work with geopackages instead. See the following example and a screenshot showing you how it looks when you open it with Qgis:

```
import fiona
from fiona.crs import from_epsg

# Whatever geometry you want
geometry = {
    "type": "Polygon",
    "coordinates": [[(10.0, 10.0), (10.0, 10.018), (10.018, 10.018), (10.018, 10.0)]]
}

schema = {
    'geometry': 'Polygon',
    'properties': {'id': 'int'},
}


file_path = "square.gpkg"
# GeoPackage
with fiona.open(
    file_path, mode='w',
    driver='GPKG',
    crs=from_epsg(4326),
    schema=schema,
    layer='eds — my supplied layer name'  
) as layer:
    layer.write({
        'geometry': geometry,
        'properties': {'id': 123},
    })
```

![image](https://github.com/Toblerity/Fiona/assets/28574023/4dc8d4fc-bfb9-4d88-b778-b4e2037d6fbc)

**Reply by nk9** [reply]
Thanks for the suggestion and the code. I want these files to be text instead of binary, so unfortunately using .gpkg won't suit my usecase this time.

**Comment by sgillies** [comment]
@nk9 there are a bunch of GeoJSON driver options that allow name and description to be serialized at the feature collection level: https://gdal.org/drivers/vector/geojson.html#layer-creation-options. Unfortunately, none of them can help you because fiona has hard-wired GeoJSON collection names. See https://github.com/Toblerity/Fiona/blob/master/fiona/collection.py#L179-L182.

This was a decision made long ago before the enhancements to the GeoJSON driver and is something that can be changed in the upcoming version 1.10.

**Reply by sgillies** [reply]
Tracking here: https://github.com/Toblerity/Fiona/issues/1351

**Reply by nk9** [reply]
This looks like just what I was hoping for. In the meantime, I guess I can write the JSON to a buffer and then add the `name` string manually before saving the buffer to a file. But yes, providing a clean way to pass this value would be great.

**Comment by jmarintur** [comment]
Hi @sgillies, nice, I didn't know about that. The times I save/export vector layers in Qgis (with supposedly gdal in the back) to geojson, the layer name option is not enabled. 

@nk9 It won't be using Fiona, but you could use for instance the geojson library (`pip install geojson`). This one allows you to write down the layer name, here's the code:

```
from geojson import Feature, FeatureCollection, Polygon

geometry = Polygon([[(10.0, 10.0), (10.0, 10.018), (10.018, 10.018), (10.018, 10.0)]])
feature = Feature(geometry=geometry)
feature_collection = FeatureCollection(features=[feature], name="eds - whatever")
# Write geojson
with open("square.geojson", "w") as f:
  f.write(str(feature_collection))
```

**Reply by nk9** [reply]
Thanks very much for the suggestion! My goal is to have a file which QGIS can open, modify, and then save without extraneous changes so that the specific lines/points/properties that _have_ changed can be reviewed in git. I actually started the project using `json.dumps()`, but the format was quite different. I've found that [Fiona produces the correct output](https://github.com/shapely/shapely/discussions/2010#discussioncomment-8726044) w.r.t. spacing and significant digits, so that's exactly what I'm looking for. Now it's just a matter of getting this name in there.

While the FeatureCollection suggestion you've made here is nice and clean, the output doesn't match what QGIS produces. So unfortunately, it won't work for my use case. I'm sorry for not explaining this at the outset, I wanted to keep the question simple!
